### ISSUE TITLE
Hyperopt / Sharpe Ratio v.s. Sortino

### ISSUE THREAD

**Issue by Prossi79** [freqtrade/freqtrade#2407]
I appreciate the option to hyperopt towards the Sharpe Ratio, what about Sortino? I recently read following intersting comment about both ratios:

"The Sortino ratio is motivated by the fact that the Sharpe ratio captures both upward and
downward volatility in its denominator. However, investors (and hedge fund managers) are generally not too bothered when we have signifcant upward volatility!
What is actually of interest from a risk management perspective is downward volatility and periods of drawdown."

Could the Sortino be implemented in the same way as the Sharpe as hyperopt option?

**Comment by xmatthias** [comment]
Depends what data Sortino uses ... but the anser is probably yes.
Have you got a good link as to how the formula for sortino is?

**Comment by hroff-1902** [comment]
@xmatthias look here, all sharpe, calmar and sortino are one-liners from returns:
https://github.com/kernc/backtesting.py/blob/a46d72a635faf06a50ad47409ebf6fa5fde81c04/backtesting/backtesting.py#L921-L923

This implementation does not do annualization of the coefficients, but I like the approach they calculated...

**Comment by hroff-1902** [comment]
Implementation from pyfolio: 
https://github.com/quantopian/pyfolio/blob/c5ebc0020f605b024edffcdcf452aca34bee8bf8/pyfolio/timeseries.py#L202

**Comment by xmatthias** [comment]
Empyrical does not have any "odd" dependencies apart from pandas-datareader which i think does not work correctly anymore - so that'll be fine to include.

Pyfolio on the other hand will not give us any benefit in this case ... since it really calls empyrical and has these ratios deprecated ... and is more usefull in interactive analysis setups.

**Comment by hroff-1902** [comment]
And even just the 'stats.py' module from empyrical. We don't need the rest.

**Comment by hroff-1902** [comment]
Another implementation from ranaroussi's quantstat (we also need only this module):
https://github.com/ranaroussi/quantstats/blob/master/quantstats/stats.py

**Comment by xmatthias** [comment]
no, i wouldn't start copy/pasting stuff. I ... maintenance of that is a huge pain (we'll always be out of date). 

There is a reasons for things like pypi ... 

The only reason for doing that is if the rest of the module has strange dependencies we absolutely don't want (which is not something i see with empyrical) or if the library is not maintained (does not mean no commits, but no responses to issues) - otherwise it's way easier to include this as a dependency and leave maintenance of this to the main library.

**Comment by hroff-1902** [comment]
Sure. As discussed, this should be a dependency, if we'll utilize this external code

**Comment by gaugau3000** [comment]
Be aware that both empyrical and pyfolio seems to be design like a Buy and Hold on an assets. For example the max_drawdown don't go below a 100% loss (of course an asset can't have a neg value) but in a backtest you can loose more than 100% :-). Don't know if a lib is more design for algo trading but not sure about these 2 libs :-)

Have a nice day

**Comment by orehunt** [comment]
@gaugau3000 I don't follow...if you have upward volatilty it means that some trades were simply more successful than others, this is mostly trend related, and I think any strategy will always profit more on uptrends and less on downtrends, so considering upward volatilty as bad is not right.

**Comment by gaugau3000** [comment]
@orehunt this topic is about a criteria that will be used to optimize a back-test in a machine learning problem. Peaks/deviation from the mean in a back-test is most of the time due to over-fitting (feel free to look and contribute in #2472 to see what i speak). A way to limit that is too take into account volatility and the Sharpe ratio is a metric that combine profit and volatility : how lucky we are :-). 

Even By the way you prefer to win a constant percentage every month or a big win and then 6 month draw-down/flat : i personally make my choice. Maybe this type of metric at a system level (with many bots uncorrelated it reduce volatility) but definitely not for me at 1 bot level.

Just my opinion you can have a different of course.

Have a nice day.

**Comment by orehunt** [comment]
@gaugau3000 I think I get it now, it probably wasn't obvious at first because I have mostly written strategies with very long pairlists, so hyperopt has an hard time overfitting. 
So In general I think if you are testing with little data sharpe is better, while if you have tons of data sortino shouldn't have problems.

**Comment by yazeed** [comment]
I've noticed that the more epochs I have hyperopt run for the higher the stoploss.

I think having Sortino would be really nice for freqtrade because you can then use it effectively to trade a long list of pairs taking drawdown effectively into consideration.

Until then, it seems we should stick to a smaller list of pairs when hyper optimizing.

**Comment by xmatthias** [comment]
> Would it be possible to reconsider this feature?

Well - we did not throw this idea away - and having more possibilities when it comes to HyperoptLoss functions can only be benefitial.
Personally - i currently lack the time to work on this - but we're always more than happy to get community contributions.

This can also be a good way to get into freqtrade code deeper - it's a relatively isolated section of the code.

I'll try to add some documentation on how to implement a new HyperoptLoss function - but it'll come down to something like copy/pasting a existing Loss function, removing the content of the `hyperopt_loss_function()` replacing it with a new implementation, changing the class name and adding that class-name to the documentaiton in `cli_options.py`.

The function needs to return lower values for better results since hyperopt minimizes this function.

**Comment by yazeed** [comment]
Yeah, I would actually code it if I knew how to know how to calculate the downside return, maybe @djacky can take a look as I'm sure he's interested in hyperopt loss methods. I'm really curious to test this one out, or even apply it in combination with the sharpe ratio loss method.

[Various Risk Methods Summarized](https://www.managedfuturesinvesting.com/8-risk-adjusted-statistics-summarized/)

**Comment by yazeed** [comment]
> > Would it be possible to reconsider this feature?
> 
> Well - we did not throw this idea away - and having more possibilities when it comes to HyperoptLoss functions can only be benefitial.
> Personally - i currently lack the time to work on this - but we're always more than happy to get community contributions.
> 
> This can also be a good way to get into freqtrade code deeper - it's a relatively isolated section of the code.
> 
> I'll try to add some documentation on how to implement a new HyperoptLoss function - but it'll come down to something like copy/pasting a existing Loss function, removing the content of the `hyperopt_loss_function()` replacing it with a new implementation, changing the class name and adding that class-name to the documentaiton in `cli_options.py`.
> 
> The function needs to return lower values for better results since hyperopt minimizes this function.
> 
> Update ~14h later:
> Let me revise this last comment - the documentation is already available [here](https://www.freqtrade.io/en/latest/advanced-hyperopt/#creating-and-using-a-custom-loss-function).

Yeah, I would actually code it if I knew how to know how to calculate the downside return, maybe @djacky can take a look as I'm sure he's interested in hyperopt loss methods. I'm really curious to test this one out, or even apply it in combination with the sharpe ratio loss method.

Good read: [Various Risk Methods Summarized](https://www.managedfuturesinvesting.com/8-risk-adjusted-statistics-summarized/)

**Comment by xmatthias** [comment]
I think downside returns are as simple as applying a filter to all negative profits.

`downside_returns = returns.loc[returns['profit_percent'] < 0]'` ... ?

**Comment by yazeed** [comment]
Yes, I was hacking this on my own last night, and reached this code, but it still needs testing to confirm the results.

```
        rfr = 0
        target = 0
        sum_daily['downside_returns'] = 0
        sum_daily.loc[sum_daily['profit_percent'] < target,
                      'downside_returns'] = sum_daily['profit_percent']**2
        expected_return = sum_daily['profit_percent'].mean()
        down_stdev = np.sqrt(sum_daily['downside_returns'].mean())

        if np.std(total_profit) != 0.0:
            sortino_ratio = (
                # sum_daily["profit_percent"].mean()
                # / sum_daily["profit_percent"].std()
                # * np.sqrt(365)
                ((expected_return - rfr)/down_stdev)
            )
        else:
            # Define high (negative) sharpe ratio to be clear that this is NOT optimal.
            sortino_ratio = -20.0

        print(sortino_ratio)
        return -sortino_ratio
```

**Comment by xmatthias** [comment]
Since we're now supporting both Sharpe, Sortino and SharpeDaily and SortinoDaily hyperopt losses, i'll close this issue.

Not sure why we forgot to lose this for over 3 months ... :laughing:
